File: leetcode/141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* front = head;
        ListNode* back = head;

        if(head == nullptr)
            return false;
        
        while( front != nullptr && back != nullptr)
        {
            back = back->next;

            if(front->next == nullptr)
                return false;
            front = front->next->next;

            if(front == back)
                return true;
        }
        return false;
    }
};

File: leetcode/141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        front = head
        back = head

        if head == None:
            return False
        
        while front != None and back != None:
            back = back.next

            if front.next == None:
                return False
            
            front = front.next.next

            if front == back:
                return True

        return False